File: Sources/XMTP/CodecRegistry.swift
//
//  CodecRegistry.swift
//
//
//  Created by Pat Nakajima on 12/22/22.
//

import Foundation

struct CodecRegistry {
	var codecs: [String: any ContentCodec] = [:]

	mutating func register(codec: any ContentCodec) {
		codecs[codec.id] = codec
	}

	func find(for contentType: ContentTypeID?) -> any ContentCodec {
		guard let contentType else {
			return TextCodec()
		}

		if let codec = codecs[contentType.id] {
			return codec
		}

		return TextCodec()
	}
}

File: library/src/main/java/org/xmtp/android/library/CodecRegistry.kt
package org.xmtp.android.library

import org.xmtp.android.library.codecs.ContentCodec
import org.xmtp.android.library.codecs.ContentTypeId
import org.xmtp.android.library.codecs.TextCodec
import org.xmtp.android.library.codecs.id

data class CodecRegistry(val codecs: MutableMap<String, ContentCodec<*>> = mutableMapOf()) {

    fun register(codec: ContentCodec<*>) {
        codecs[codec.contentType.id] = codec
    }

    fun find(contentType: ContentTypeId?): ContentCodec<*> {
        contentType?.let {
            val codec = codecs[it.id]
            if (codec != null) {
                return codec
            }
        }
        return TextCodec()
    }

    fun findFromId(contentTypeString: String): ContentCodec<*> {
        for ((_, codec) in codecs) {
            if (codec.contentType.id == contentTypeString) {
                return codec
            }
        }
        return TextCodec()
    }
}
